### PR TITLE
Add --fineGrainedHashExternalRepos

### PR DESCRIPTION
**Pull request by tgeng**
Hi friends at Tinder! First thank you for making this awesome project! I work for Pinterest and we are evaluating adopting this over our in-house solution.

One issue I am seeing is that bazel-diff seems to treat the entire external repo as one opaque blob. For example, if one uses this with `maven_install`, any single third party dependency change ends up trigger a global rebuild.

This PR adds a flag that makes bazel-diff generate fine-grained hash for the explicitly mentioned external repos. This way, bazel-diff is smarter when handling external repos like `@maven` in a monorepo setting.

Usage:
pass `--fineGrainedHashExternalRepos maven1,maven2` to `generate-hashes` command and fine-grained hashes are generated for targets in the listed external repos.

Note: This is still a draft and hence I have not added any tests to this yet. I would like to get some input on this approach before that. I have tested this with a mono repo with >4M lines of Java code and it seems to be working as expected.

### CODE REVIEW

File: cli/src/main/kotlin/com/bazel_diff/cli/GenerateHashesCommand.kt [hunk 1]
```diff
@@ -64,6 +65,14 @@ class GenerateHashesCommand : Callable<Int> {
     )
     var bazelCommandOptions: List<String> = emptyList()
 
+    @CommandLine.Option(
+        names = ["--fineGrainedHashExternalRepos"],
```
Comment (tinder-maxwellelliott) [inline cli/src/main/kotlin/com/bazel_diff/cli/GenerateHashesCommand.kt:69]:
```suggestion
        names = ["--fine-grained-hash-external-repos"],
```

File: cli/src/main/kotlin/com/bazel_diff/bazel/BazelClient.kt [hunk 2]
```diff
@@ -4,16 +4,17 @@ import com.bazel_diff.log.Logger
...
-class BazelClient : KoinComponent {
+class BazelClient(private val fineGrainedHashExternalRepos: Set<String>) : KoinComponent {
     private val logger: Logger by inject()
     private val queryService: BazelQueryService by inject()
 
     suspend fun queryAllTargets(): List<BazelTarget> {
         val queryEpoch = Calendar.getInstance().getTimeInMillis()
-        val targets = queryService.query("'//external:all-targets' + '//...:all-targets'")
+
+        val query = listOf("//external:all-targets", "//...:all-targets") + fineGrainedHashExternalRepos.map { "@$it//...:all-targets" }
```
Comment (tinder-maxwellelliott) [inline cli/src/main/kotlin/com/bazel_diff/bazel/BazelClient.kt:16]: I like it!

**Review by tinder-maxwellelliott — COMMENTED** [review]
Really like the approach! Minor comments

File: cli/src/main/kotlin/com/bazel_diff/cli/GenerateHashesCommand.kt (hunk 1 above)
Comment (tgeng) [reply cli/src/main/kotlin/com/bazel_diff/cli/GenerateHashesCommand.kt]: Thanks! Updated and also added an integration test covering this flag.

**Comment by tinder-maxwellelliott** [comment]
@tgeng I am sorry, I am now realizing that `fineGrainedHashExternalRepos` matches the existing arg style. Can you go back to that?

**Comment by tgeng** [comment]
> @tgeng I am sorry, I am now realizing that `fineGrainedHashExternalRepos` matches the existing arg style. Can you go back to that?

Np! Pushed another patch.

**Comment by tinder-maxwellelliott** [comment]
@tgeng Thank you!

**Comment by tgeng** [comment]
Thanks for the quick response! When can I expect another release? (I can build our own, but we still want to make it clean eventually 😉 )

**Comment by tinder-maxwellelliott** [comment]
> Thanks for the quick response! When can I expect another release? (I can build our own, but we still want to make it clean by using the official build eventually 😉 )

I will upload it now!

**Pull request merged**